clang-tools-extra/clang-tidy/misc/UseInternalLinkageCheck.cpp

```diff
@@ -0,0 +1,78 @@
+//===--- UseInternalLinkageCheck.cpp - clang-tidy--------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "UseInternalLinkageCheck.h"
+#include "../utils/FileExtensionsUtils.h"
+#include "clang/AST/Decl.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/ASTMatchers/ASTMatchersMacros.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Basic/Specifiers.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::misc {
+
+namespace {
+
+AST_MATCHER(Decl, isFirstDecl) { return Node.isFirstDecl(); }
+
+AST_MATCHER_P(Decl, isInMainFile, FileExtensionsSet, HeaderFileExtensions) {
+  return llvm::all_of(Node.redecls(), [&](const Decl *D) {
+    SourceManager &SM = Finder->getASTContext().getSourceManager();
+    const SourceLocation L = D->getLocation();
+    return SM.isInMainFile(L) &&
+           !utils::isSpellingLocInHeaderFile(L, SM, HeaderFileExtensions);
+  });
```

Comment:
should be:

```
#source files or main file if its not a header.
return utils::isSpellingLocInFile(L, SM, SourceFileExtensions) ||
          (SM.isInMainFile(L) && !utils::isSpellingLocInFile(L, SM, HeaderFileExtensions));
```

isSpellingLocInHeaderFile should be renamed into isSpellingLocInFile, that "Header" is not needed in those functions.

Reply:
It will match more cases which I cannot understand the code style.
e.g.
```c++
/// a.cpp
void f(); // decl-1

/// b.cpp
#include <a.cpp>
void f() {} // decl-2
```
The current version will ignore this cases because decl-1 is not in main file.
But the suggestion version will catch this cases because both file is source fill extensions.

Actually I don't understand and didn't meet this code style, so I keep conservative to avoid false positive.

Reply:
"because decl-1 is not in main file" but thats still source file and thats a point.

Take into account 2 use cases:
1. user want to run check on a header files, in such case header file will be a main file, and check wont work.
2. user want to run check on unity cpp, where there is single .cpp file generated by cmake that include other cpp files and those files include header files.
In both cases user should be able to do that. 

You already got "allOf(isFirstDecl(), isInMainFile(HeaderFileExtensions)," and thats fine. so check should work only if first declaration is in source file (no additional header) or when declaration and definition is in header file, in such case adding static/anonymous namespace is also expected (but only if user explicitly want to run check on headers - by messing up with compile commands json for example).

Reply:
I got the point. Our main disagreement is about

> when declaration and definition is in header file, in such case adding static/anonymous namespace is also expected

In my opinion, we should not emit any wrong for the definition in header file. Add static in header file is dangerous and cause potential bug.

e.g.
```c++
/// a.hpp
PREFIX int f() {
  static int a = 1;
  return a++;
}
void b();

/// a.cpp
#include "a.hpp"
#include <iostream>

int main() {
  std::cout << __func__ << " " << f() << "\n";
  b();
  std::cout << __func__ << " " << f() << "\n";
  b();
}

/// b.cpp
#include "a.hpp"
#include <iostream>

void b() { std::cout << __func__ << " " << f() << "\n"; }
```

if `PREFIX` is `inline`, result is `1 2 3 4`.
if `PREFIX` is `static`, result is `1 2 1 2`.
if `PREFIX` is `inline static`, result is `1 2 1 2`.

if `PREFIX` is `template <int>`, result is `1 2 3 4`.
if `PREFIX` is `template <int> static`, result is `1 2 1 2`.

Reply:
I'm fine for lack of fixes in header files.

Reply:
this still won't work for "unity files" (includes .cpp)

Reply:
what is the expected behavior for unity files? use static or not? I am a little bit confused.

Reply:
As a summary
If main file include non-header file (*.inc, *.cpp, etc..), this check should treat it as main file.
Then we just need to check there are no `redecl` in header file.

Reply:
yes, that FirstDecl stuff will work. just note that we got config for header extensions and source extensions. better would be to use source. as those .inc could be part of header files also.

Reply:
I think we can recursive to find the real file of non-header file. Tracking the include chain and find the first file which includes this decl.

Reply:
No need to do recursive, using (current) first declaration should also work.